internal/controllers/rest_controller.go

```diff
@@ -169,31 +195,122 @@ func (rc RESTController) UserInfo(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+// GetToken returns data from the Vault.
+func (rc RESTController) GetToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	data, err := rc.provisionTokenService.Get(path)
+	if err != nil {
+		rc.badRequest(w, err)
+		return
+	}
+
+	json.NewEncoder(w).Encode(data)
+	w.WriteHeader(http.StatusOK)
+}
+
+// ListTokens returns a data list from the Vault.
+func (rc RESTController) ListTokens(c *gin.Context) {
+	r, w := c.Request, c.Writer
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	data, err := rc.provisionTokenService.List(path)
+	if err != nil {
+		rc.badRequest(w, err)
+		return
+	}
+
+	json.NewEncoder(w).Encode(data)
+	w.WriteHeader(http.StatusOK)
+}
+
+// CreateToken stores data into Vault.
+func (rc RESTController) CreateToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
+
+	var newProvisionToken models.ProvisionTokenRequest
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		rc.internalServerError(w, err)
+		return
+	}
+	json.Unmarshal(reqBody, &newProvisionToken)
+
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	data, err := rc.provisionTokenService.Create(path, newProvisionToken)
+	if err != nil {
+		rc.badRequest(w, err)
+		return
+	}
+
+	expiration := time.Duration(newProvisionToken.Expiration) * time.Hour
+	jsonBody, err := rc.oidcService.Create(data, expiration)
+	if err != nil {
+		rc.internalServerError(w, err)
 		return
 	}
 
-	rc.deleteSessionCookie(w, "access_token")
+	json.NewEncoder(w).Encode(jsonBody)
+	w.WriteHeader(http.StatusCreated)
+}
+
+// DeleteToken removes a data from the Vault.
+func (rc RESTController) DeleteToken(c *gin.Context) {
+	r, w := c.Request, c.Writer
+	resource := strings.Split(r.URL.Path, "/")[2]
+	uuid := c.Params.ByName("uuid")
+	path := fmt.Sprintf("%s/data/%s", resource, uuid)
+
+	tokenToBeDeleted := c.Query("token_id")
+	if tokenToBeDeleted == "" {
+		rc.badRequest(w, errors.New("no token was provided in the query string"))
+		return
+	}
+
+	err := rc.provisionTokenService.Delete(path, tokenToBeDeleted)
+	if err != nil {
+		rc.badRequest(w, err)
+	}
+
 	w.WriteHeader(http.StatusOK)
 }
 
 // getCredentials is responsible for simply getting credentials.
-func (rc RESTController) getCredentials(r *http.Request) (string, error) {
+func (rc RESTController) getCredentials(r *http.Request) (string, string, error) {
 	rc.logger.Info("Get request credentials...")
 	for _, auth := range rc.credentials {
 		token, err := auth.Extract(r)
 		if authErrors.IsTokenNotFound(err) {
 			continue
 		}
-		// credentials sent.
 		rc.logger.Debug(token)
-		return token, nil
+		// credentials sent.
+		return token, reflect.TypeOf(auth).Elem().Name(), nil
```

Comment:
Why do we need to use reflection here and what that information means?

Reply:
to identify what the request is: cookie or authorization header (bearer).